Issue/woomob 843 woo posbarcodes barcodes for pos scanner setup

malinajirka opened the pull request:
<!-- Remember about a good descriptive title. -->

<!-- Id number of the Linear issue this PR addresses, e.g., WOOMOB-373. -->
WOOMOB-868

### Description
<!-- Take the time to write a good summary. Why is it needed? What does it do? When fixing bugs try to avoid just writing “See original issue” – clarify what the problem was and how you’ve fixed it. -->

I ended up making more changes than I originally planned. However, vast majority of the changed lines are test and move of the UiStates definitions. The PR should hopefully still be relatively easy to follow. Let me know if not and I'll split it.

This PR refactors the WooPosScanningSetupViewModel
- Moves state definitions into the state class.
- Extracts the next/previous step logic into a separate class + adds a bunch of tests to verify all the combinations.
- Replaces strings for UI with R.string.[id] to avoid having to use companion objects helping keep the state definitions clean.
- Adds some missing tests to the VM.

All these changes have two main goals:
1. It'll be easy to have a different flow for different readers.
2. The VM was growing and splitting helped keeping even the test classes easier to get oriented in.

### Testing information
<!-- This is your opportunity to break out individual scenarios that need testing (when necessary) and/or include a checklist for the reviewer to go through. Consider documenting the following from your own completed testing: devices used, alternate workflows, edge cases, affected areas, critical flows, areas not tested, and any remaining unknowns. Provide feedback on this new section of the PR template through Sept 30, 2024 to Apps Quality; additional context here: p91TBi-b8z-p2#comment-12036 -->

Requires full re-testing of the Barcode Setup flows - having said that, I assume we'll do a test plan testing so I don't think it's necessary to spent more than 10 minutes on the testing.

### The tests that have been performed
<!-- To give the reviewer an idea of what could be missed in terms of testing -->

I've tried to test all the states of the setup flow that came to my mind.

### Images/gif
<!-- Include before and after images or gifs when appropriate. -->
N/A

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

malinajirka committed:
Introduce different configuration for different scanners

malinajirka committed:
Move barcode handling logic to Navigator

malinajirka committed:
Simplify ScanningSetup state handling for different readers

malinajirka committed:
Add tests for ScannerNavigator and ScanningSetupVM

malinajirka committed:
Add WooPos prefix

malinajirka committed:
Fix device selection state

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

malinajirka committed:
Remove manual handling of device selection step

malinajirka committed:
Move initial state handling to the navigator

malinajirka committed:
Move remaining navigation logic to navigator

malinajirka committed:
Move remaining navigation logic to navigator

malinajirka committed:
Handle steps that aren't part of the main flow

malinajirka committed:
Fix detekt

malinajirka committed:
Merge branch 'trunk' into issue/woomob-843-woo-posbarcodes-barcodes-for-pos-scanner-setup

malinajirka commented on WooCommerce/src/test/kotlin/com/woocommerce/android/ui/woopos/home/scanningsetup/WooPosScanningSetupViewModelTest.kt:
```diff
@@ -34,299 +35,305 @@ class WooPosScanningSetupViewModelTest {
-        return WooPosScanningSetupViewModel(resourceProvider)
-    }
-
-    private fun setupMockResourceProvider() {
```
These are not longer needed since the UIState contains just the id not the string itself.

malinajirka committed:
Remove unused code

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/scanningsetup/WooPosScannerSetupNavigator.kt:
```diff
@@ -0,0 +1,79 @@
+    }
+
+    fun getPreviousStep(device: BarcodeReaderDevice, currentStep: ScanningSetupStep): ScanningSetupStep {
+        // Handle special branching steps that aren't in the main sequences
```
np: probably redundant

malinajirka replied:
I think this one is valuable since it might not be clear that there is a main sequence and then branches. I'll keep it.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/scanningsetup/WooPosScannerSetupNavigator.kt:
```diff
@@ -0,0 +1,79 @@
+                val stepSequence = ScannerConfigurations.getStepSequence(device)
+                val currentIndex = stepSequence.indexOf(currentStep)
+
+                if (currentIndex <= 0) {
```
We should probably crash here if `currentIndex`, the same as in `getNextStep`? -1 - means not found.

malinajirka replied:
Good point, fixed in 48eab7a693f058fb26dd7e2775417bab2d81dccc

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/scanningsetup/WooPosScannerSetupNavigator.kt:
```diff
@@ -0,0 +1,79 @@
+        return when (currentStep) {
+            is ScanningSetupStep.TestYourScannerTimeout,
+            is ScanningSetupStep.TestYourScannerScanFailed -> {
+                getPreviousStep(device, ScanningSetupStep.TestYourScanner)
```
Shell it be on navigation back 

* TestYourScannerTimeout -> TestYourScanner
* TestYourScannerScanFailed -> TestYourScannerTimeout

?

Now it opens the step which is before `TestYourScanner`

malinajirka replied:
Good question, and I'm not sure what is expected. I was working with the assumption that all `TestYourScanner...` are the same step, so if we want to go back, we go back to Pairing. 

Thinking about this more, I'd say the Failed state shouldn't have back at all - only retry => if we read an invalid barcode, we are "sure" the scanner isn't correctly setup. The TestYourScannerTimeout should go back to pairing as it does now, since it's still the same step with the same code as TestYourScanner. Wdyt?

kidinov replied:
I agree with your thinking except that:
* As we implemented it, it's now visually a new step. We play animation, and the size of the dialog changes so it feels and looks like just a new step visually, the same as the others
* While we mostly sure that failed step means something is wrong with the scanner, there is still a small chance that it was an actual keyboard used by mistake

All together, considering that these are rather edge cases, I'd keep it simple implementation and flow-wise and always just navigate back

```
    fun getPreviousStep(device: BarcodeReaderDevice, currentStep: ScanningSetupStep): ScanningSetupStep {
        val stepSequence = ScannerConfigurations.getStepSequence(device)
        val currentIndex = stepSequence.indexOf(currentStep)
        return stepSequence[currentIndex - 1]
    }
```

malinajirka replied:
I believe `val currentIndex = stepSequence.indexOf(currentStep)` will return -1 if the step is one of the side-branches, right?

> As we implemented it, it's now visually a new step. We play animation, and the size of the dialog changes so it feels and looks like just a new step visually, the same as the others
While we mostly sure that failed step means something is wrong with the scanner, there is still a small chance that it was an actual keyboard used by mistake

Good points! Perhaps going to TestYourScanner makes the most sense.

kidinov replied:
> I believe val currentIndex = stepSequence.indexOf(currentStep) will return -1 if the step is one of the side-branches, right?

Oh yes, it will. I sort of liked how it was before - linked list for back navigation. Stack also would be nice. Having said that, I don't know if it's worth a change now

malinajirka replied:
I kind of liked the original approach as well - however, having the definition of all the steps at the compile time IMO outweighs the downsides. I wish we could somehow elegantly include the side-branches in the definition of the flow, that would be cleanest. I didn't come up with a reasonable solution though since the side-branches are used only when certain conditions are met. There will always need to be some business logic behind it => this logic IMO belongs to the VM not the navigator or the state definition.

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/scanningsetup/WooPosScanningSetupState.kt:
```diff
@@ -18,92 +19,231 @@ data class WooPosScanningSetupState(
-            val bluetoothSettingsButtonText: String,
-            override val previousStep: ScanningSetupStep
-        ) : ScanningSetupStep()
+            @DrawableRes val iconRes: Int = R.drawable.ic_woopos_bluetooth_settings
```
Do you expect this icon to be specific per reader? It looks like all the resources that are not specific are hardcoded, so maybe it also should be an object?

malinajirka replied:
I missed this, fixed in 9f72bb0c9a09f30fb309601aca1cd3c59f2cf202

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/scanningsetup/WooPosScanningSetupDialog.kt:
```diff
@@ -165,21 +165,21 @@ fun WooPosScanningSetupDialog(
-                        title = step.title,
-                        message = step.message,
+                        title = stringResource(step.titleRes),
+                        message = stringResource(step.messageRes),
```
It's a parameterised string here, therefore Int probably cannot be used. Initially, they were used everywhere for this reason - I thought that there would be many places like that where we mention the device name.

<img width="555" height="360" alt="image" src="https://github.com/user-attachments/assets/b72bbac1-fd81-4e22-b62d-160d4b651e6c" />

malinajirka replied:
Good catch! Not sure how I could have missed this - I'll try to come up with a solution that will still allow us to avoid companion objects.

malinajirka replied:
What do you think about this approach @kidinov ?
<img width="1102" height="206" alt="Screenshot 2025-07-16 at 9 36 01" src="https://github.com/user-attachments/assets/3c31dc8e-847f-4157-a9ba-2fc73ddb043f" />

kidinov replied:
I guess it's fine, although tbh I am not a big fan of it. It feels like it's better if the VM exposes ready to show data to the view

malinajirka replied:
I know what you mean, but technically I believe it's an anti-pattern => the view might have Themed composable or resources might change during a configuration change so getting resources should be IMO done in the view layer. Having said that, it's more in the sphere of theory since both approaches are likely fine and the downsides are edge-cases not worth worrying about. Having said all that, in this case, since you are not strongly against, I'll opt for the UIString because I want to avoid having to pass `resourceProvider` to all the ScanningSetupSteps - that will save a bunch of boilerplate.

malinajirka committed:
Crash if index out of bounds

kidinov reviewed (commented):
Looks and tests well! Thanks for the improvement!

I left a couple of comments; please take a look!

malinajirka committed:
Share BT settings icon in UI state

malinajirka committed:
Fix tests

malinajirka committed:
Use parametrized string

malinajirka committed:
Use correct reader name

malinajirka committed:
Navigate back to TestYourScanner from side branches

kidinov commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/woopos/home/scanningsetup/WooPosScanningSetupState.kt:
```diff
@@ -18,92 +20,240 @@ data class WooPosScanningSetupState(
+                BarcodeReaderDevice.STAR_BSH_20B -> listOf(
+                    ScanningSetupStep.DeviceSelection,
+                    ScanningSetupStep.ScannerHIDModeSetup(qrCodeImageRes = R.drawable.ic_barcode),
+                    ScanningSetupStep.PairYourScanner(deviceName = device.displayNameRes),
```
‼️ `ScannerPairModeSetup` is missing for this reader. Expected?

malinajirka replied:
Yes, this is expected since we have all-in-one barcode for this reader and that's displayed on the ScannerHidModeSetup => this was actually one of the reasons I realized each scanner will need a different flow.

malinajirka committed:
Update tests

malinajirka commented:
Thanks @kidinov for the review! I've implemented the suggestions and fixed the issues you found. It should be all good now, but definitely worth re-reviewing. Thanks!

malinajirka committed:
Merge branch 'trunk' into issue/woomob-843-woo-posbarcodes-barcodes-for-pos-scanner-setup

kidinov reviewed (approved):
LGTM!
